packages/transport-webtransport/src/index.ts

```diff
@@ -52,60 +64,122 @@ class WebTransportTransport implements Transport {
-      serverCertificateHashes: certhashes.map(certhash => ({
-        algorithm: 'sha-256',
-        value: certhash.digest
-      }))
-    })
-    wt.closed.catch((error: Error) => {
-      log.error('WebTransport transport closed due to:', error)
-    })
-    await wt.ready
-
     if (remotePeer == null) {
       throw new Error('Need a target peerid')
     }
 
-    if (!await this.authenticateWebTransport(wt, localPeer, remotePeer, certhashes)) {
-      throw new Error('Failed to authenticate webtransport')
+    if (certhashes.length === 0) {
+      throw new Error('Expected multiaddr to contain certhashes')
     }
 
-    const maConn: MultiaddrConnection = {
-      close: async (options?: AbortOptions) => {
-        log('Closing webtransport')
-        await wt.close()
-      },
-      abort: (err: Error) => {
-        log('Aborting webtransport with err:', err)
-        wt.close()
-      },
-      remoteAddr: ma,
-      timeline: {
-        open: Date.now()
-      },
-      // This connection is never used directly since webtransport supports native streams.
-      ...inertDuplex()
-    }
+    let abortListener: (() => void) | undefined
+    let maConn: MultiaddrConnection | undefined
 
-    wt.closed.catch((err: Error) => {
-      log.error('WebTransport connection closed:', err)
-      // This is how we specify the connection is closed and shouldn't be used.
-      maConn.timeline.close = Date.now()
-    })
+    let cleanUpWTSession: (closeInfo?: WebTransportCloseInfo) => void = () => {}
 
     try {
+      let closed = false
+      const wt = new WebTransport(`${url}/.well-known/libp2p-webtransport?type=noise`, {
+        serverCertificateHashes: certhashes.map(certhash => ({
+          algorithm: 'sha-256',
+          value: certhash.digest
+        }))
+      })
+
+      cleanUpWTSession = (closeInfo?: WebTransportCloseInfo) => {
+        try {
+          if (maConn != null) {
+            if (maConn.timeline.close != null) {
+              // already closed session
+              return
+            }
+
+            // This is how we specify the connection is closed and shouldn't be used.
+            maConn.timeline.close = Date.now()
+          }
+
+          if (closed) {
+            // already closed session
+            return
+          }
+
+          wt.close(closeInfo)
+        } catch (err) {
+          log.error('error closing wt session', err)
+        } finally {
+          closed = true
+        }
+      }
+
+      // this promise resolves/throws when the session is closed or failed to init
+      wt.closed
+        .then(async () => {
+          await maConn?.close()
+        })
+        .catch((err: Error) => {
+          log.error('error on remote wt session close', err)
+          maConn?.abort(err)
+        })
+        .finally(() => {
+          // if we never got as far as creating the maConn, just clean up the session
+          if (maConn == null) {
+            cleanUpWTSession()
+          }
+        })
```

Comment:
I believe this should cover the case where `await wt.ready` throws, but from my testing, I'm not confident the `.closed` catch always handles errors thrown from `wt.ready`.